# SpliceS: Fast scanner for conserved splice sites in gene homologs across species #


SpliceS is a tool to identify splice sites conserved across multiple species, facialiating primer design in transcript homologs of novel species. Developed by the [Center for Research Informatics (CRI) bioinformatics core](http://cri.uchicago.edu/?page_id=1185) at the University of Chicago. 

The pipeline is implemented in 

* [Perl](https://www.perl.org/) 


# Quick Start #

To gain the first look into how to tool works, we provide sample data and scripts for a quick start demo.

In directory example/

* **KiSS1.hamster_RNAseq.alignment.fasta** *Transcript sequence alignment*
* **run_example.sh** *Example commands to run the tool*
* **KiSS1.hamster_RNAseq.alignment.fasta.out files** *Output files generated by the commands*
* **run_example.log** *Log file after running the commands*

To regenerate the output files, try to run 
 
```
#!bash
cd example
./run_example.sh

```

# Documentation #

Please see the [Wiki](https://github.com/riyuebao/SpliceS/wiki) for full documentation.

# Contact #

Questions and comments? Please contact Riyue Bao at \<rbao at uchicago dot edu\>, or post on [Issues](https://github.com/riyuebao/SpliceS/issues).

# Release #

* **Version 0.1.0** *2016-06-28*

# TODO #

* Add alignment function 
* Add vsualization
