# SpliceS: Fast scanner for conserved splice sites in gene homologs across species #


SpliceS is a tool to identify splice sites conserved across multiple species, facialiating primer design in transcript homologs of novel species. Developed by the [Center for Research Informatics (CRI) bioinformatics core](http://cri.uchicago.edu/?page_id=1185) at the University of Chicago. 

The pipeline is implemented in 

* [Perl](https://www.perl.org/) 

<img src="http://i.imgur.com/rhhS3tQ.png" width=1000/>

# Download #

To download SpliceS, please use git to download the most recent version.

```
$ git clone https://github.com/riyuebao/SpliceS.git
```

# System Requirement #

* SpilceS requires `perl 5.18.1` or higher. 
* To visualize the sequences in Fasta format or generate alignment: [MEGA](http://www.megasoftware.net/) 
* To quickly map sequence to the genome and visualize the gene model: [BLAT](https://genome.ucsc.edu/cgi-bin/hgBlat?command=start).

# Usage #

For a description of available command-line options and their defaults, run:

```
$ ./SpliceS.pl

--------------------------------------------------------------------------------
Copyright (c) 2016 Center for Research Informatics, University of Chicago

Usage: SpliceS.pl [OPTIONS]
Vesion: 0.1.0

Options:
 [-i|--input]     : Input sequence file. Fasta format.
                    Sequence header format must contain three fields,
                    separated by [|].  No space allowed.
                    For example: >gene_symbol|transcript_accessionID|species
                    [species] must be
                    rn or rn6; mm or mm10; hg or hg38.
             Note : All three homolog sequences are required (rn, mm, hg).
 [-o|--output]    : Output file prefix. Default is "$input.out".
 [-rn|--rat]      : Rat RefSeq gene annotation. refFlat format.
 [-mm|--mouse]    : Mouse RefSeq gene annotation. refFlat format.
 [-hg|--human]    : Human RefSeq gene annotation. refFlat format.
             Note : All three annotation files are required (rn, mm, hg).

Three output files will be generated:
 $output.seq.txt : Non-gapped transcript sequence with splice site labeled.
 $output.gene_model.txt: Gene model information of known species.
 $output.splice_pos.txt: Splice position information of the gene homologs.

Example: SpliceS.pl \
         -i KiSS1.hamster_RNAseq.alignment.fasta \
         -rn ../data/rn6.refFlat.txt \
         -mm ../data/mm10.refFlat.txt \
         -hg ../data/hg38.refFlat.txt

Contact: Riyue Bao <rbao@uchicago.edu>
--------------------------------------------------------------------------------

```

# Quick Start #

To gain the first look into how to tool works, we provide sample data and scripts for a quick start demo.

In directory example/

* **KiSS1.hamster_RNAseq.alignment.fasta** *Transcript sequence alignment*
* **run_example.sh** *Example commands to run the tool*
* **KiSS1.hamster_RNAseq.alignment.fasta.out files** *Output files generated by the commands*
* **run_example.log** *Log file after running the commands*

To regenerate the output files, run:
 
```
cd example
./run_example.sh

```



# Contact #

Questions and comments? Please contact Riyue Bao at \<rbao at uchicago dot edu\>, or post on [Issues](https://github.com/riyuebao/SpliceS/issues).

# Release #

* **Version 0.1.0** *2016-06-28*

# TODO #

* Add alignment function 
* Add vsualization
